File: snippets/hmac.py
import requests
import requests.auth
import base64
import time
import json
import binascii
from hashlib import sha256
from urllib.parse import urlparse


class QtradeAuth(requests.auth.AuthBase):

    def __init__(self, key):
        self.key_id, self.key = key.split(":")

    def __call__(self, req):
        # modify and return the request
        timestamp = str(int(time.time()))
        url_obj = urlparse(req.url)

        request_details = req.method + "\n"
        """
        request_details = 'GET
        '
        """
        uri = url_obj.path
        # uri = "/v1/user/orders"
        if url_obj.query:
            uri += "?" + url_obj.query
            # uri = "/v1/user/orders?open=true"
        request_details += uri + "\n"
        """
        request_details = 'GET
        /v1/user/orders?open=true
        '
        """
        request_details += timestamp + "\n"
        """
        request_details = 'GET
        /v1/user/orders?open=true
        1573604427
        '
        """
        if req.body:
            # this request has no body, so this code isn't run in this example
            if isinstance(req.body, str):
                request_details += req.body + "\n"
            else:
                request_details += req.body.decode('utf8') + "\n"
        else:
            request_details += "\n"
            """
            request_details = 'GET
            /v1/user/orders?open=true
            1573604427

            '
            """
        request_details += self.key
        """
        request_details = 'GET
        /v1/user/orders?open=true
        1573604427

        1111111111111111111111111111111111111111111111111111111111111111'
        """
        hsh = sha256(request_details.encode("utf8")).digest()
        signature = base64.b64encode(hsh)
        req.headers.update({
            "Authorization": "HMAC-SHA256 {}:{}".format(self.key_id, signature.decode("utf8")),
            "HMAC-Timestamp": timestamp
        })
        return req


# Create a session object to make repeated API calls easy!
api = requests.Session()
# Create an authenticator with your API key
api.auth = QtradeAuth("1:1111111111111111111111111111111111111111111111111111111111111111")

# Make a call to API
res = api.get('https://api.qtrade.io/v1/user/orders', params={'open': "true"}).json()
print(res)

# Generated headers are:
{'Accept': '*/*',
 'Accept-Encoding': 'gzip, deflate',
 'Authorization': 'HMAC-SHA256 1:111111111111111111111111111111111111111111',
 'Connection': 'keep-alive',
 'HMAC-Timestamp': '1574589817',
 'User-Agent': 'python-requests/2.18.4'}


# Generated body is:
{'data': {'orders': [{'created_at': '2019-11-12T22:42:15.643486Z',
                      'id': 8932525,
                      'market_amount': '1.10428011',
                      'market_amount_remaining': '1.10428011',
                      'market_id': 1,
                      'open': True,
                      'order_type': 'sell_limit',
                      'price': '0.0083759',
                      'trades': None},
                     {'created_at': '2019-11-12T22:42:14.713136Z',
                      'id': 8932523,
                      'market_amount': '0.92023342',
                      'market_amount_remaining': '0.92023342',
                      'market_id': 1,
                      'open': True,
                      'order_type': 'sell_limit',
                      'price': '0.00788731',
                      'trades': None},
                     {'base_amount': '0.00433166',
                      'created_at': '2019-11-12T22:42:08.51142Z',
                      'id': 8932503,
                      'market_amount': '14862.44638875',
                      'market_amount_remaining': '14862.44638875',
                      'market_id': 36,
                      'open': True,
                      'order_type': 'buy_limit',
                      'price': '0.00000029',
                      'trades': None}]}}

File: snippets/hmac.js
var crypto = require('crypto');
var https = require('https');

class QtradeAuth {
  constructor(key) {
    [this.keyID, this.key] = key.split(":");
    this.host = 'api.qtrade.io';
  }
  req(method, endpoint, body='', handleData) {
    let timestamp = Math.floor(Date.now() / 1000);

    // Create hmac sig
    let sig_text = method + "\n";
    sig_text += endpoint + "\n";
    sig_text += timestamp + "\n";
    sig_text += body + "\n";
    sig_text += this.key;
    var hash = crypto.createHash('sha256').update(sig_text, 'utf8').digest().toString('base64');

    var opt = {
      host: this.host,
      path: endpoint,
      method: method,
      headers: {
        'Content-Type': 'application/json; charset=utf-8',
        'Authorization': 'HMAC-SHA256 ' + this.keyID + ':' + hash,
        "HMAC-Timestamp": timestamp,
      }
    };

    var output = '';
    let req = https.request(opt, (res) => {
      res.setEncoding('utf8');
      res.on('data', (chunk) => {
        output += chunk;
      });
      res.on('end', () => {
        handleData(JSON.parse(output));
      })
    });
    req.write(body);
    req.end();
  }
  get(endpoint, handleData) {
    return this.req("GET", endpoint, undefined, handleData);
  }
  post(endpoint, body, handleData) {
    return this.req("POST", endpoint, body, handleData);
  }
}

let api = new QtradeAuth("1:1111111111111111111111111111111111111111111111111111111111111111");

// GET request
api.get("/v1/user/me", (resp) => {
  console.log(resp);
});

// POST request
payload = {
  "amount": "1",
  "price": "0.01",
  "market_id": 1,
};
api.post("/v1/user/sell_limit", JSON.stringify(payload), (resp) => {
  console.log(resp);
});


// Generated headers are:
{
  data: {
    user: {
      id: 1111111,
      email: 'hugh@example.com',
      lname: 'Jass',
      fname: 'Hugh',
      verified_email: true,
      tfa_enabled: false,
      withdraw_limit: 0,
      verification: 'none',
      can_withdraw: true,
      can_login: true,
      can_trade: true,
      referral_code: '6W56QFFVIIJ2',
      email_addresses: [Array]
    }
  }
}


// Generated body is:
{
  data: {
    order: {
      id: 8932314,
      market_amount: '1',
      market_amount_remaining: '1',
      created_at: '2018-13-12T23:18:01.070552Z',
      price: '0.01',
      order_type: 'sell_limit',
      market_id: 1,
      open: true,
      trades: []
    }
  }
}
